src/utils/src/vm_memory.rs

```diff
@@ -444,4 +690,147 @@ mod tests {
                 .unwrap();
         }
     }
+
+    #[test]
+    fn test_read_volatile() {
+        let test_cases = [
+            ([1u8, 2].as_ref(), [1u8, 2, 0, 0, 0]),
+            ([1, 2, 3, 4].as_ref(), [1, 2, 3, 4, 0]),
+            // ensure we don't have a buffer overrun
+            ([5, 6, 7, 8, 9].as_ref(), [5, 6, 7, 8, 0]),
+        ];
+
+        for (mut input, output) in test_cases {
+            // ---- Test ReadVolatile for &[u8] ----
+            //
+            // Test read_volatile for &[u8] works
+            let mut memory = vec![0u8; 5];
+
+            assert_eq!(
+                input
+                    .read_volatile(&mut VolatileSlice::from(&mut memory[..4]))
+                    .unwrap(),
+                input.len().min(4)
+            );
+            assert_eq!(&memory, &output);
+
+            // Test read_exact_volatile for &[u8] works
+            let mut memory = vec![0u8; 5];
+            let result = input.read_exact_volatile(&mut VolatileSlice::from(&mut memory[..4]));
+
+            // read_exact fails if there are not enough bytes in input to completely fill
+            // memory[..4]
+            if input.len() < 4 {
+                result.unwrap_err();
```

Comment:
Would it add value checking against the exact error variant?

Reply:
More value checking is always good, I'll add it :)